lib/spack/spack/test/architecture.py

```diff
@@ -169,3 +169,22 @@ def test_target_container_semantic(cpu_flag, target_name):
 def test_arch_spec_container_semantic(item, architecture_str):
     architecture = spack.spec.ArchSpec(architecture_str)
     assert item in architecture
+
+
+@pytest.mark.parametrize('compiler,real_version,target_str,expected_flags', [
+    (spack.spec.CompilerSpec('gcc@9.2.0'), None, 'haswell',
+     '-march=haswell -mtune=haswell'),
+    # Check that custom string versions are accepted
+    (spack.spec.CompilerSpec('gcc@foo'), '9.2.0', 'icelake',
+     '-march=icelake-client -mtune=icelake-client'),
+])
```

Comment:
add one with a version and a suffix to make sure that works (e.g., `gcc@4.4.9-special`)

Reply:
There are fixes in #13250 to treat suffixes correctly. Can we merge that one first and rebase this?

Reply:
@alalazo: done!  This needs a rebase now.

Reply:
> add one with a version and a suffix to make sure that works (e.g., gcc@4.4.9-special)

I think we need a bit of discussion to check what does it mean "it works". Currently suffixes are part of what identifies a compiler and in the current state `gcc@4.4.9-special` won't be considered the same as `gcc@4.4.9`. This was done for `clang` since the version numbers between official `llvm` and Apple are unrelated and distinguished just by a suffix in the version.

Things standing like this I see three possible ways forward:
1. We keep things as they are in both Spack and `archspec.cpu` and we document that if a user wants to customize the version of his compiler he shouldn't leave the number in
2. We keep a list of "officially" allowed suffixes in Spack (I think this will be just `-apple` for the moment) and if a suffix is not in the list we consider it a custom version
3. We implement @becker33 suggestion to store also the version number in `compilers.yaml`. At that point we won't use `compiler.version` for the match on optimization flags but `compiler.version_number`. Apple would use a version number with the `-apple` suffix

3 is probably the cleanest solution, but I don't have clear yet if it requires some special care to be backward compatible. 2 might be a good alternative and would be faster to implement. It's compatible with 3 and will be undone as soon as we implement it. 1 is super fast but it might cause inconveniences if users are employing custom versions like `gcc@4.4.9-special` since it requires to change that.

Reply:
Seeing that we have already projects for v0.14 and v0.15 I'll go on and implement 2 if there are no objections. 3 should probably go in one of the next versions.

Reply:
@tgamblin @becker33 I went with solution 2 in fc96281. If that's not what we want the commit can be easily reverted. If the semantic is fine we can merge this PR as is and I can create an issue to implement 3 above at a later point.

Reply:
I agree with this approach, and I agree that we can add the caching later

Reply:
me too